twister: support parallel coverage with lcov >=2.0

tari opened the pull request:
lcov 2.0 added support for processing coverage data in parallel, which provides a large speedup when processing many files, at the cost of some additional overhead. When running the Chrome EC tests with coverage, parallel reporting on a 36C72T machine reduces the time spent generating coverage reports by 40 minutes (from approximately 1 hour to 20 minutes total runtime), at the cost of about 3x greater CPU time overall (assumed to be overhead for parallel processing, likely from spawning much larger numbers of subprocesses).

The level of lcov parallelism is taken from the --jobs option passed to twister, allowing lcov to choose if unspecified.

tari committed:
twister: support parallel coverage with lcov >=2.0

lcov 2.0 added support for processing coverage data in parallel, which
provides a large speedup when processing many files, at the cost of some
additional overhead. When running the Chrome EC tests with coverage,
parallel reporting on a 36C72T machine reduces the time spent generating
coverage reports by 40 minutes (from approximately 1 hour to 20 minutes
total runtime), at the cost of about 3x greater CPU time overall
(assumed to be overhead for parallel processing, likely from spawning
much larger numbers of subprocesses).

The level of lcov parallelism is taken from the --jobs option passed to
twister, allowing lcov to choose if unspecified.

Signed-off-by: Peter Marheine <pmarheine@chromium.org>

golowanow commented:
@tari, just 2 quick notes from my side, could you please:
1) split your changes into at least 2 commits: one for all the refactoring changes, and another for the lcov parallel execution option introduced.
2) make the lcov parallel execution mode aligned with the twister option `--jobs`, so lcov should use not more than that max. number of CPU threads.

tari committed:
twister: move lcov version handling to helpers

This moves handling of changed lcov arguments in lcov 2.0 into helper
functions, significantly simplifying Lcov._generate() to reduce the
visual noise of extra arguments that are required but unimportant to the
task at hand.

Signed-off-by: Peter Marheine <pmarheine@chromium.org>

tari commented:
> 1. split your changes into at least 2 commits: one for all the refactoring changes, and another for the lcov parallel execution option introduced.

Done.

> 2. make the lcov parallel execution mode aligned with the twister option `--jobs`, so lcov should use not more than that max. number of CPU threads.

Good idea, done.
